A Semaphore for Parallelizable Reentrancy Protection (#5283)
ethereum/EIPs: EIPS/eip-5283.md

## Specification

Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address `0x0A`. When `Semaphore` is called, if the caller address is present more than once in the call stack, the contract behaves as if the first instruction had been a `REVERT`, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 100, which is consumed independently of the call result.

## Rationale

The only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.

> Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM. The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits because most transactions use the same [EIP-20](./eip-20.md) contracts.

Reviewer: These paragraphs read more like Motivation than Rationale. The Motivation section describes why the EIP as a whole needs to exist, while the Rationale section should describe individual technical choices made within the EIP.

> The address `0x0A` is the next one available within the rage defined by [EIP-1352](./eip-1352).

Reviewer: rage?

Reviewer 2: 💢

Author: :)

---
eip: 5283
title: Semaphore for Parallelizable Reentrancy Guard
description: A Precompile-based parallelizable reentrancy protection using the call stack
author: Sergio D. Lerner (@SergioDemianLerner)>
status: Draft
type: Standards Track
category: Core
created: 2022-07-17

> ---

Reviewer 2: ```suggestion
requires: 20, 1109, 1153, 1283, 1352, 2046
---
```

Author: Should an EIP referenced be in the "requires" list even if it doesn't need to be final for this EIP to be finalized ? For example, should competing proposals be referenced in "requires" ?

Reviewer 3: You generally shouldn't include competing proposals in your EIP.  An EIP should stand on its own as much as possible, and external references should be to final EIPs (or EIPs that are expected to be final before your EIP makes it to final).  Keep in mind, the EIP process is for standardization and isn't an appropriate place to argue for/against this standard vs that standard.  Such material is better suited for a blog, article, docs site, discussions-to link, forum thread, etc.

My usual recommendation to authors is to remove all references to other EIPs that aren't required reading for understanding this standard, and if you want to mention something in the rationale as "we thought about doing X but decided against it because Y" then just briefly describe X rather than linking to a non-final EIP.

Keep in mind, EIPs can be completely/radically changed prior to becoming final so if you write your EIP and reference a non-final EIP then it is possible that the content you referenced may no longer make sense, which is why we don't allow non-final references in final EIPs.

## Abstract

This EIP proposes adding a precompiled contract that provides a semaphore function for creating a new type of reentrancy protection guard (RPG). This function aims to replace the typical RPG based on modifying a contract storage cell. The benefit is that the precompile-based RPG does not write to storage, and therefore it enables contracts to be forward-compatible with all designs that provide fine-grained (i.e. cell level) parallelization for the multi-threaded execution of EVM transactions. 

## Motivation

> The typical smart contract RPG uses a contract storage cell. The algorithm is simple: the code checks that a storage cell is 0 (or any other predefined constant) on entry, aborting if not, and then sets it to 1. After executing the required code, it resets the cell back to 0 before exiting. This is the algorithm implemented in OpenZeppelin's [ReentrancyGuard](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol). The algorithm results in a read-write pattern on the RPG's storage cell. This pattern prevents the parallelization of the execution of the smart contract for all known designs that try to provide fine-grained parallelization (detecting conflicts at the storage cell level). 

Reviewer 2: ```suggestion
The typical smart contract RPG uses a contract storage cell. The algorithm is simple: the code checks that a storage cell is 0 (or any other predefined constant) on entry, aborting if not, and then sets it to 1. After executing the required code, it resets the cell back to 0 before exiting. This is the algorithm implemented in OpenZeppelin's ReentrancyGuard. The algorithm results in a read-write pattern on the RPG's storage cell. This pattern prevents the parallelization of the execution of the smart contract for all known designs that try to provide fine-grained parallelization (detecting conflicts at the storage cell level). 
```

## Specification

Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address `0x0A`. When `Semaphore` is called, if the caller address is present more than once in the call stack, the contract behaves as if the first instruction had been a `REVERT`, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 100, which is consumed independently of the call result.

## Rationale

The only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.

> Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM (i.e. see [RSKIP-144](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP144.md)). The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits, because most transactions use the same ERC20 contracts.

Reviewer 2: ```suggestion
Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM. The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits because most transactions use the same [EIP-20](./eip-20.md) contracts.
```

     * Calling a `nonReentrant` function from another `nonReentrant`
     * function is supported.      
     */
    modifier nonReentrant() {
        _nonReentrantBefore();
        _;
    }

    function _nonReentrantBefore() private {
    	assembly {
            if iszero(staticcall(1000,SemaphoreAddress,  0, 0, 0, 0)) {
                revert(0, 0)
            }
        }
    }
}
```

### Parallellizable storage-based RPGs
 
The only way to paralellize prexistent contracts that are using the storage RPG construction is that the VM automatically detects that a storage variable is used for the RPG, and proves that is works as required. This requires static code analysys. This is difficult to implement in consensus for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the RPG, meaning that some execution paths do not alter the RPG, which may complicate proving. Therefore this proposal aims to protect future contracts and let them be parallelizable, rather than to paralellize already deployed ones.

### Alternatives

There are alternative designs to implement RPGs on the EVM:

> 1.	[EIP-1153: Transient storage opcodes](https://eips.ethereum.org/EIPS/eip-1153) or [`SSTORE_COUNT`](https://github.com/ethereum/EIPs/issues/119). The first does reduce the cost of an RPG, but does not address the problem of parallelization. The second enables execution parallelization but it is way more complex as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value. Also the `SSTORE` count is not currently tracked by EVM. Also, implementing new opcodes is discuraged if the same functionality can be implemented using precompiles.

Reviewer 3: Instead of linking to these, you should briefly describe their behavior.  This EIP won't be able to move to final until all of the EIPs it references are also final, and since this EIP is "competing" with these you certainly wouldn't want to be blocked by them not being included.

Author: Ok, I'll describe them. But I can still mention the IP numbers right ?

Reviewer 3: Not unless you want to be blocked by them.  Until an EIP is final, it is completely mutable.  If you refer to EIP-X today and then your EIP becomes final but EIP-X is not, then it is possible that EIP-X will be radically changed before it is finally merged as final and now your EIP doesn't make sense because it is referring to a document that is not the same as when you wrote the reference.

Author: Understood!

2. A new `LOCKCALL` opcode is introduced that works quite similar to `STATICALL` but only blocks storage writes in the caller contract. This results in cheaper RPG, but has two drawbacks: it doesn't allow some contract functions to be free of the RPG, and it intruduces a new opcode which requires modifying compilers, debuggers and static analysys tools.

### Gas cost
A gast cost of 100 represents a worst case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.  Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in totall).

> The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](https://eips.ethereum.org/EIPS/eip-1283). Using the `Semaphore` precompile as a reentrancy check would currently costs 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost, and therefore discourages its use,  it is still much lower than the pre-EIP1283 cost. If a reduction in precompile call cost (such as [EIP-1109](https://eips.ethereum.org/EIPS/eip-1109) or [EIP-2046](https://eips.ethereum.org/EIPS/eip-2046)) is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use the precompile-based RPG, it is suggested that this EIP is implemented together with EIP-2046.

Reviewer 2: ```suggestion
The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost (such as [EIP-1109](./eip-1109.md) or [EIP-2046](./eip-2046.md)) is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with EIP-2046.
```

## Backwards Compatibility
This change is a hard-fork and therefore all full nodes must be updated.

## Test Cases

```
contract Test is ReentrancyGuard2 {
    function second() external nonReentrant {
    }
    function first() external nonReentrant {
        this.second();
    }

> }

Author: Yul is ok?

```
A call to second() directly from a transaction does not revert, but a call to first() does revert.

> ## Reference Implementation
> None

Reviewer 2: ```suggestion
```

Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM. The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits because most transactions use the same [EIP-20](./eip-20.md) contracts.

To summarize, the only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.

## Specification

> Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address `0x0A`. When `Semaphore` is called, if the caller address is present more than once in the call stack, the contract behaves as if the first instruction had been a `REVERT`, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 100, which is consumed independently of the call result.

Reviewer 4: A precompiled contract usually takes time (month to years) to converge to a consensus. I suggest use a placeholder for `0x0A` and finalize it before the doc becomes finalized or before a hardfork EIP specifying it

## Rationale

> The address `0x0A` is the next one available within the range defined by [EIP-1352](./eip-1352).

Reviewer 4: Same here

> ### Sample usage

Reviewer 4: IMHO this is more suitable to be in the reference implementation (of usage)

     * Calling a `nonReentrant` function from another `nonReentrant`
     * function is supported.      
     */
    modifier nonReentrant() {
        _nonReentrantBefore();
        _;
    }

    function _nonReentrantBefore() private {
    	assembly {
            if iszero(staticcall(1000,SemaphoreAddress,  0, 0, 0, 0)) {
                revert(0, 0)
            }
        }
    }
}
```

### Parallellizable storage-based RPGs
 
The only way to paralellize prexistent contracts that are using the storage RPG construction is that the VM automatically detects that a storage variable is used for the RPG, and proves that is works as required. This requires static code analysys. This is difficult to implement in consensus for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the RPG, meaning that some execution paths do not alter the RPG, which may complicate proving. Therefore this proposal aims to protect future contracts and let them be parallelizable, rather than to paralellize already deployed ones.

### Alternatives

There are alternative designs to implement RPGs on the EVM:

> 1.	Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes also enable fine-grained parallelization. 

Reviewer 4: ```suggestion
1. Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes also enable fine-grained parallelization. 
```

2. An opcode `SSTORE_COUNT` that retrieves the number of `SSTORE` instructions executed. It enables also fine-grained execution parallelization, but `SSTORE_COUNT` is much more complex to use correctly as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value.
3. A new `LOCKCALL` opcode that works similar to `STATICALL` but only blocks storage writes in the caller contract. This results in cheaper RPG, but it doesn't allow some contract functions to be free of the RPG.

> All these alternate proposals have the downside that they create new opcodes, and this is discuraged if the same functionality can be implemented with the same gas cost using precompiles. A new opcode requires modifying compilers, debuggers and static analysys tools.

Reviewer 4: ```suggestion
All these alternative proposals have the downside that they create new opcodes, and this is discouraged if the same functionality can be implemented with the same gas cost using precompiles. A new opcode requires modifying compilers, debuggers and static analysis tools.
```

### Gas cost

A gas cost of 100 represents a worst-case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.

Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in total).

> The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with a reduction in precompile calls cost.

Reviewer 4: [Optional] Could you further provide your suggestion of how much the gas cost would you suggest?

## Backwards Compatibility

This change requires a hard fork and therefore all full nodes must be updated.

## Test Cases

> ```

Reviewer 2: ```suggestion
```solidity
```
Nit.

To summarize, the only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.

## Specification

Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address `0x0A`. When `Semaphore` is called, if the caller address is present more than once in the call stack, the contract behaves as if the first instruction had been a `REVERT`, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 100, which is consumed independently of the call result.

## Rationale

The address `0x0A` is the next one available within the range defined by [EIP-1352](./eip-1352).

### Sample usage

> ```

Reviewer 2: ```suggestion
```solidity
```

    	assembly {
            if iszero(staticcall(1000,SemaphoreAddress,  0, 0, 0, 0)) {
                revert(0, 0)
            }
        }
    }
}
```

### Parallellizable storage-based RPGs
 
The only way to paralellize prexistent contracts that are using the storage RPG construction is that the VM automatically detects that a storage variable is used for the RPG, and proves that is works as required. This requires static code analysys. This is difficult to implement in consensus for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the RPG, meaning that some execution paths do not alter the RPG, which may complicate proving. Therefore this proposal aims to protect future contracts and let them be parallelizable, rather than to paralellize already deployed ones.

### Alternatives

There are alternative designs to implement RPGs on the EVM:

1.	Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes also enable fine-grained parallelization. 

> 

Reviewer 2: ```suggestion
```

2. An opcode `SSTORE_COUNT` that retrieves the number of `SSTORE` instructions executed. It enables also fine-grained execution parallelization, but `SSTORE_COUNT` is much more complex to use correctly as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value.

> 

Reviewer 2: ```suggestion
```

The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with a reduction in precompile calls cost.

## Backwards Compatibility

This change requires a hard fork and therefore all full nodes must be updated.

## Test Cases

```
contract Test is ReentrancyGuard2 {
    function second() external nonReentrant {
    }
    function first() external nonReentrant {
        this.second();
    }
}
```

> A call to second() directly from a transaction does not revert, but a call to first() does revert.

Reviewer 2: ```suggestion
A call to `second()` directly from a transaction does not revert, but a call to `first()` does revert.
```

## Security Considerations

> No security risks has been identified related to this change.

Reviewer 2: ```suggestion
No security risks have been identified related to this change.
```

## Copyright

> Copyright and related rights waived via [CC0](../LICENSE.md).

Reviewer 2: ```suggestion

Copyright and related rights waived via [CC0](../LICENSE.md).
```

> ---

Reviewer 2: Missing discussions-to.

eip: 5283
title: Semaphore for Reentrancy Protection
description: A Precompile-based parallelizable reentrancy protection using the call stack
author: Sergio D. Lerner (@SergioDemianLerner)
status: Draft

> type: Standards Track
> category: Core

Reviewer 2: ```suggestion
type: Standards Track
category: ERC
```
```suggestion
type: Informational
```
This EIP is not a core EIP. Please pick one of these (I'm inclined towards the latter).

Reviewer 3: Unless I'm reading this EIP wrong, it *is* a Core EIP as it introduces a precompile to the EVM.

Reviewer 2: Disregard my previous comment.

created: 2022-07-17
requires: 20, 1283, 1352
---

> 

Reviewer 2: ```suggestion
```

Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM. The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits because most transactions use the same [EIP-20](./eip-20.md) contracts.

To summarize, the only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.

> 

Reviewer 2: ```suggestion
```

## Specification

Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address `0x0A`. When `Semaphore` is called, if the caller address is present more than once in the call stack, the contract behaves as if the first instruction had been a `REVERT`, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 100, which is consumed independently of the call result.

> 

Reviewer 2: ```suggestion
```

## Rationale

> 

Reviewer 2: ```suggestion
```

The address `0x0A` is the next one available within the range defined by [EIP-1352](./eip-1352).

### Sample usage

```
pragma solidity ^0.8.0;

abstract contract ReentrancyGuard2 {

    uint8 constant SemaphoreAddress = 0x0A;
    /**
     * @dev Prevents a contract from calling itself, directly or indirectly.
     * Calling a `nonReentrant` function from another `nonReentrant`
     * function is supported.      
     */
    modifier nonReentrant() {
        _nonReentrantBefore();
        _;
    }

    function _nonReentrantBefore() private {
    	assembly {
            if iszero(staticcall(1000,SemaphoreAddress,  0, 0, 0, 0)) {
                revert(0, 0)
            }
        }
    }
}
```

> 

Reviewer 2: ```suggestion
```

### Parallellizable storage-based RPGs
 
The only way to paralellize prexistent contracts that are using the storage RPG construction is that the VM automatically detects that a storage variable is used for the RPG, and proves that is works as required. This requires static code analysys. This is difficult to implement in consensus for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the RPG, meaning that some execution paths do not alter the RPG, which may complicate proving. Therefore this proposal aims to protect future contracts and let them be parallelizable, rather than to paralellize already deployed ones.

### Alternatives

There are alternative designs to implement RPGs on the EVM:

> 1.	Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes do not address the problem of fine-grained parallelization. 

Reviewer 5: why doesn't this help with parallelization? transactions that use any amount of transient storage can be executed in parallel, even if they use the same slot

Reviewer 5: is this about parallel execution for a single transaction?

Author: My mistake! I misplaced that sentence when rewritting the paragraph according to the last review.

Reviewer 5: so i understand correctly, the transient storage eip provides the same benefits as this eip, and is more general, i.e. not limited to reentrancy locks? would be helpful to have a `discussions-to` url in the header early so there's a better place to discuss, but also think it should be explained in the alternate or rationale how this eip is different

Author: yes, transient opcodes are better from the EVM "purity" standpoint. But I haven't heard of any use for the transient opcodes apart from mutexes. I will add the discussions tag, where we should be chatting.

2. An opcode `SSTORE_COUNT` that retrieves the number of `SSTORE` instructions executed. It enables also fine-grained execution parallelization, but `SSTORE_COUNT` is much more complex to use correctly as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value.

3. A new `LOCKCALL` opcode that works similar to `STATICALL` but only blocks storage writes in the caller contract. This results in cheaper RPG, but it doesn't allow some contract functions to be free of the RPG.

> 

Reviewer 2: ```suggestion
```

All these alternate proposals have the downside that they create new opcodes, and this is discuraged if the same functionality can be implemented with the same gas cost using precompiles. A new opcode requires modifying compilers, debuggers and static analysys tools.

> 

Reviewer 2: ```suggestion
```

### Gas cost

> A gast cost of 100 represents a worst case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.  Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in totall).

Reviewer 2: ```suggestion

A gas cost of 100 represents a worst-case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.

Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in total).
```

The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with a reduction in precompile calls cost.

> 

Reviewer 2: ```suggestion
```

## Backwards Compatibility

> This change is a hard-fork and therefore all full nodes must be updated.

Reviewer 2: ```suggestion

This change requires a hard fork and therefore all full nodes must be updated.
```

## Test Cases

```
contract Test is ReentrancyGuard2 {
    function second() external nonReentrant {
    }
    function first() external nonReentrant {
        this.second();
    }
}
```

> A call to second() directly from a transaction does not revert, but a call to first() does revert.

Reviewer 2: ```suggestion

A call to second() directly from a transaction does not revert, but a call to first() does revert.
```

> 

Reviewer 2: ```suggestion
```

## Security Considerations

> No security risks have been identified related to this change.

Reviewer 2: ```suggestion
Needs discussion.
```

Author: No security risk has been identified in recent discussions.

## Security Considerations

> No security risks has been identified related to this change.

Reviewer 2: ```suggestion
No security risks have been identified related to this change.
```
